### PR TITLE
Bug fix for on exiting a remote context-manager with exception

### PR DESCRIPTION
**Pull request by koreno**
When exiting a remote context-manager due to an exception, rpyc will attempt to pass the `exc, typ, tb` to the remote `__exit__` of the context-manager. Python then rightfully complains that the `tb` isn't a proper `traceback` object:
```
E           _get_exception_class.<locals>.Derived: throw() third argument must be a traceback object
E           
E           ========= Remote Traceback (1) =========
E           Traceback (most recent call last):
E             File "/home/ofer/Sources/wekapp/deps/rpyc/rpyc/core/protocol.py", line 312, in _dispatch_request
E               res = self._HANDLERS[handler](self, *args)
E             File "/home/ofer/Sources/wekapp/deps/rpyc/rpyc/core/protocol.py", line 562, in _handle_callattr
E               return self._handle_getattr(oid, name)(*args, **dict(kwargs))
E             File "/usr/lib/python3.5/contextlib.py", line 77, in __exit__
E               self.gen.throw(type, value, traceback)
E           TypeError: throw() third argument must be a traceback object
```

### CODE REVIEW

File: rpyc/utils/teleportation.py [hunk 1]
```diff
@@ -117,7 +117,10 @@ def _import_codetup(codetup):
 
 def import_function(functup):
     name, modname, defaults, codetup = functup
-    mod = __import__(modname, None, None, "*")
+    try:
```
Comment (coldfix) [inline rpyc/utils/teleportation.py]: This change seems unrelated to the PR topic. Also, I think it makes generally no sense to use  `__main__` as fallback here (or does it?) and letting the import fail will result in better error messages. I will probably cherry-pick around this commit.

File: rpyc/core/protocol.py [hunk 2]
```diff
@@ -560,6 +560,15 @@ def _handle_setattr(self, oid, name, value):
         return self._access_attr(oid, name, (value,), "_rpyc_setattr", "allow_setattr", setattr)
     def _handle_callattr(self, oid, name, args, kwargs):
         return self._handle_getattr(oid, name)(*args, **dict(kwargs))
+    def _handle_ctxexit(self, oid, exc):
+        if exc:
```
Comment (coldfix) [inline rpyc/core/protocol.py]: What's the reason for the `try...except` here? To support context managers that require real traceback object?

The traceback generated here is not the one corresponding to the original exception, so I'm almost thinking it might be better to let context managers fail that require a traceback. I'd tend to remove the whole `if...else` block and simply pass `self._handle_getattr(oid, "__exit__")(exc, type(exc), tb)` instead. CMs that require tracebacks are probably more of an edge-case anyway. What do you think?

File: tests/test_context_managers.py [hunk 3]
```diff
@@ -6,14 +6,19 @@
```
Comment (coldfix) [inline tests/test_context_managers.py]: These should be instance variables. Otherwise you can't call the `exposed_context` method twice (or you have to deal with resetting them after each call but that's worse..).

File: tests/test_context_managers.py [hunk 4]
```diff
@@ -24,18 +29,30 @@ def setUp(self):
...
             self.assertFalse(on_context_exit)
             print( "got past context enter" )
             self.assertEqual(x, 20)
             print( "got past x=20" )
+        self.assertFalse(on_context_exc)
         self.assertTrue(on_context_exit)
         print( "got past on_context_exit" )
 
+    def test_context_exception(self):
+        with self.assertRaises(AssertionError):
```
Comment (coldfix) [inline tests/test_context_managers.py]: There is two things wrong here:
- having multiple asserts within the `with` block means we don't know which one generated the exception
- checking against `AssertionError` confuses the boundaries between test failures and desired behaviour

I think the test should look like this:

```
with self.conn.root.context(3):
    raise ValueError()
self.assertTrue(on_context_enter)
self.assertTrue(on_context_error)
self.assertTrue(on_context_exit)
```

(without the `print`, don't ask me why they are in the other test..)

and remove the `raise` line from the context manager.

File: rpyc/utils/teleportation.py (hunk 1 above)
Comment (koreno) [inline rpyc/utils/teleportation.py]: This commit seem to have tagged along unintentionally. you may indeed drop it

File: rpyc/core/protocol.py (hunk 2 above)
Comment (koreno) [inline rpyc/core/protocol.py]: Indeed it's to provide a valid traceback object. My only fear is CMs that are written naively, like:
```
def __exit__(self, *args):
    if args[0]:
        log_traceback(args[2])
```
In other words, expecting there to always be a traceback object if an exception object was passed.

File: tests/test_context_managers.py (hunk 3 above)
Comment (koreno) [inline tests/test_context_managers.py]: I didn't want to change it too much, so I've added a `reset` method

File: tests/test_context_managers.py (hunk 4 above)
Comment (coldfix) [reply tests/test_context_managers.py]: To clarify, is this PR a fix for locally raising an exception within the context of a remote context manager, or a fix for remotely reraising it from within the `__exit__` handler of the context manager?

If the first one is the case, this test should be simplified as I suggested.

**Comment by coldfix** [comment]
Hi @koreno, I will merge your PR if you clarify the following:

(I had asked this before, but it was attached to a outdated diff, so I'm not sure you saw my query)

----

To clarify, is this PR a fix for locally raising an exception within the context of a remote context manager, or a fix for remotely reraising it from within the `__exit__` handler of the context manager?

If the first one is the case, this test should be simplified as I suggested.

----

where the suggestion was:

There is two things wrong here:
- having multiple asserts within the `with` block means we don't know which one generated the exception
- checking against `AssertionError` confuses the boundaries between test failures and desired behaviour

I think the test should look like this:

```
with self.conn.root.context(3):
    raise ValueError()
self.assertTrue(on_context_enter)
self.assertTrue(on_context_error)
self.assertTrue(on_context_exit)
```

(without the `print`, don't ask me why they are in the other test..)

and remove the `raise` line from the context manager.

**Comment by koreno** [comment]
> To clarify, is this PR a fix for locally raising an exception within the context of a remote context manager, or a fix for remotely reraising it from within the __exit__ handler of the context manager?
Indeed!

**Comment by coldfix** [comment]
> The test I've added, in its final version, indeed matches what you've suggested.

It doesn't. There are several assertions within the `with` block, line 56-58. Maybe you forgot to push?

**Comment by koreno** [comment]
Hm. I see I misunderstood your comment.
> having multiple asserts within the with block means we don't know which one generated the exception
Why is that? a python traceback clearly indicates from which line an exception got thrown.

> checking against AssertionError confuses the boundaries between test failures and desired behaviour
The test `assertRaises(MyException)` (and raises a `MyException`), so I don't see what would be confusing.

**Comment by coldfix** [comment]
Ok fine.

**Comment by coldfix** [comment]
Thanks!

**Comment by koreno** [comment]
Thank you too!
